test/test_broker.hpp

```diff
@@ -1124,47 +1249,25 @@ class test_broker {
 
         auto& ep = *spep;
 
+        auto& idx = sessions_.get<tag_con>();
+        auto it  = idx.find(spep);
+        BOOST_ASSERT(it != idx.end());
+
         // For each subscription that this connection has
         // Compare against the list of topic filters, and remove
         // the subscription if the topic filter is in the list.
-        {
-            auto & idx = subs_online_.get<tag_con>();
-            auto const& range = boost::make_iterator_range(idx.equal_range(spep));
-            for(auto it = range.begin(); it != range.end(); ) {
-                bool match = false;
-                for(auto const& topic_filter : topic_filters) {
-                    if(it->topic_filter == topic_filter) {
-                        /*
-                         * Advance the iterator using the return from erase.
-                         * The returned it may be equal to the next topic filter in
-                         * the list, so we continue comparing. When the loop
-                         * finishes, however, we can't blindly advance the
-                         * iterator, because that might result in skipping
-                         * an iterator that hasn't been checked. Further, we
-                         * can't only rely on the inner loop to advance the
-                         * iterator because we might not have a match, and
-                         * thus would infinitely loop on the current iterator.
-                         */
-                        MQTT_LOG("mqtt_broker", trace)
-                            << MQTT_ADD_VALUE(address, spep.get())
-                            << "subs_online_.erase() " << "topic_filter:" << topic_filter;
-                        it = idx.erase(it);
-                        match = true;
-                        break;
-                    }
-                }
-                if( ! match)
-                {
-                    std::advance(it, 1);
-                }
-            }
-        }
-
-        for(auto const& item : boost::make_iterator_range( subs_online_.get<tag_con>().equal_range(spep))) {
-            (void)item;
-            for(auto const& topic_filter : topic_filters) {
-                (void)topic_filter;
-                BOOST_ASSERT(item.topic_filter != topic_filter);
+        for (auto const& topic_filter : topic_filters) {
+            auto handle = subs_map_.lookup(topic_filter);
+            if (handle) {
+                idx.modify(
+                    it,
+                    [&](auto& e) {
+                        e.handles.erase(handle.value());
+                        e.clean();
```

Comment:
You call clean on a single unsubscribe, but this causes all subscriptions to be removed. If you have two subscriptions, and you unsubscribe one, the clean causes all subscriptions to be removed

Reply:
Thank you for debugging! Fixed.